runner/src/stream-handler/worker.ts

```diff
@@ -2,69 +2,152 @@ import { isMainThread, parentPort, workerData } from 'worker_threads';
 import promClient from 'prom-client';
 
 import Indexer from '../indexer';
-import RedisClient from '../redis-client';
+import RedisClient, { type StreamType } from '../redis-client';
 import { METRICS } from '../metrics';
+import type { Block } from '@near-lake/primitives';
+import LakeClient from '../lake-client';
 
 if (isMainThread) {
   throw new Error('Worker should not be run on main thread');
 }
-
-const indexer = new Indexer('mainnet');
-const redisClient = new RedisClient();
+interface QueueMessage {
+  block: Block
+  streamMessageId: string
+}
+type PrefetchQueue = Array<Promise<QueueMessage>>;
+
+interface WorkerContext {
+  redisClient: RedisClient
+  lakeClient: LakeClient
+  queue: PrefetchQueue
+  streamKey: string
+  streamType: StreamType
+}
 
 const sleep = async (ms: number): Promise<void> => { await new Promise((resolve) => setTimeout(resolve, ms)); };
 
 void (async function main () {
   const { streamKey } = workerData;
+  const redisClient = new RedisClient();
+  const workerContext: WorkerContext = {
+    redisClient,
+    lakeClient: new LakeClient(),
+    queue: [],
+    streamKey,
+    streamType: redisClient.getStreamType(streamKey),
+  };
 
   console.log('Started processing stream: ', streamKey);
 
-  let indexerName = '';
-  const streamType = redisClient.getStreamType(streamKey);
-  const isHistorical = streamType === 'historical';
+  await handleStream(workerContext, streamKey);
+})();
+
+async function handleStream (workerContext: WorkerContext, streamKey: string): Promise<void> {
+  void blockQueueProducer(workerContext, streamKey);
+  void blockQueueConsumer(workerContext, streamKey);
+}
+
+function incrementId (id: string): string {
+  const [main, sequence] = id.split('-');
+  return `${Number(main) + 1}-${sequence}`;
+}
+
+async function waitForQueueSpace (workerContext: WorkerContext): Promise<number> {
+  const HISTORICAL_BATCH_SIZE = 100;
+  return await new Promise<number>((resolve) => {
+    const intervalId = setInterval(() => {
+      const preFetchCount = HISTORICAL_BATCH_SIZE - workerContext.queue.length;
+      if (preFetchCount > 0) {
+        clearInterval(intervalId);
+        resolve(preFetchCount);
+      }
+    }, 100);
+  });
+}
+
+async function blockQueueProducer (workerContext: WorkerContext, streamKey: string): Promise<void> {
+  let streamMessageStartId = '0';
 
   while (true) {
-    try {
-      const startTime = performance.now();
+    const preFetchCount = await waitForQueueSpace(workerContext);
```

Comment:
```suggestion
    const preFetchCount = HISTORICAL_BATCH_SIZE - workerContext.queue.length;
    if (preFetchCount <= 0) {
      await sleep(100);
      continue;
    }
```

Seems simpler just to add this here rather than creating an interval?